synapse/config/_util.py

```diff
@@ -64,3 +65,30 @@ def json_error_to_config_error(
         else:
             path.append(str(p))
     return ConfigError(e.message, path)
+
+
+Model = TypeVar("Model", bound=BaseModel)
+
+
+def validate_instance_map_config(
+    config: Any,
+    model_type: TypeAlias,
```

Comment:
Oh, but I also think you need `Model = TypeVar("Model", bound=Type[BaseModel])` above.

Reply:
That is a permutation I had not tried, I did not know a `bound` would except a `Type[]` directly(in respect I should have guessed it could with Python's flexibility). That being said, the errors mypy gives are the same as the other's I've tried:

```log
synapse/config/_util.py:87: error: Variable "model_type" is not valid as a type  [valid-type]
synapse/config/_util.py:87: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
synapse/config/workers.py:219: error: Incompatible types in assignment (expression has type "Dict[str, Type[InstanceLocationConfig]]", variable has type "Dict[str, InstanceLocationConfig]")  [assignment]
```
which is what led me to finding `TypeAlias`, which worked even if not for the correct reasons. [`parse_obj_as()`](https://github.com/pydantic/pydantic/blob/be7a58210566f2cdf36f0543fda59bcc3967b15a/pydantic/tools.py#L19) source did not yield inspiration for me.

(Probably should rename the `model_type` argument to just `model` at this point)

Reply:
Mypy and Pydantic are using the annotation system here with slightly different semantics and purposes. Go with

```python
Model = TypeVar("Model", bound=BaseModel)

def validate_instance_map_config(
    config: Any,
    model_type: Type[Model],
) -> Dict[str, Model]:
    """Parse `config` as a mapping from strings to a given `Model` type.
    Args:
        config: The configuration data to check
        model_type: The BaseModel to validate and parse against.
    Returns:
        Fully validated and parsed Dict[str, Model].
    Raises:
        ConfigError, if given improper input.
    """
    try:
        # type-ignore: mypy doesn't like constructing `Dict[str, model_type]` because
        # `model_type` is a runtime variable. Pydantic is fine with this.
        instances = parse_obj_as(Dict[str, model_type], config)  # type: ignore[valid-type]
    except ValidationError as e:
        raise ConfigError(str(e)) from e
    return instances
```

The key thing is that the annotations on the function are correct. We can see that this is the case because mypy can deduce the type of the return value:

```
        # A map from instance name to host/port of their HTTP replication endpoint.
        self.instance_map = validate_instance_map_config(
            config.get("instance_map", {}),
            InstanceLocationConfig,
        )
        reveal_type(self.instance_map)
        # synapse/config/workers.py:221: note: Revealed type is "builtins.dict[builtins.str, synapse.config.workers.InstanceLocationConfig]"

```

Reply:
I think that's what I had when I first put it together to test, based on the function you linked above(without the `# type: ignore`, of course). When mypy complained, I thought I had messed up something. I believe I understand now, thank you for clarifying the why.

Fixed in [cfe048e](https://github.com/matrix-org/synapse/pull/15431/commits/cfe048ecd677bcfb04edc5cb7020c8f9e9e9af57)